Critical flows: Merchant → Checkout → Can adjust T&S and Privacy Policy options

nielslange opened the pull request:
Partially fixes #5793

### Notes

This PR adds an e2e test to verify that a merchant can adjust the T&C and Privacy Policy links on the checkout block.

### Testing

#### Automated Tests
* [x] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [x] E2E tests

### Manual Testing

1. Check out this PR.
2. Start Docker Desktop.
3. Run `npm run wp-env destroy && npm run wp-env start` to start a fresh e2e environment.
4. Run `npm run test:e2e -- checkout-terms` to run only this test case.
5. Run `npm run test:e2e` to run the whole test suite.

nielslange committed:
Add test to check if T&C and Privacy links are visible

tarunvijwani commented on .wp-env.json:
```diff
@@ -3,6 +3,7 @@
 	"plugins": [
 		"https://downloads.wordpress.org/plugin/woocommerce.latest-stable.zip",
 		"https://github.com/WP-API/Basic-Auth/archive/master.zip",
+		"https://github.com/nielslange/woo-test-helper/archive/refs/heads/trunk.zip",
```
Should we move this zip to our repository?
Like under woocommerce-gutenberg-products-block/tests/utils/?

tarunvijwani commented on tests/e2e/specs/merchant/checkout-terms.test.js:
```diff
@@ -0,0 +1,102 @@
+describe( 'Merchant → Checkout → Can adjust T&S and Privacy Policy options', () => {
+	beforeAll( async () => {
+		await shopper.goToShop();
+		await page.goto( 'http://localhost:8889/?setup_terms_and_privacy' );
```
Instead of `http://localhost:8889/`, we can use `BASE_URL`. We'd be able to avoid using hardcoded values this way.

tarunvijwani commented on tests/e2e/specs/merchant/checkout-terms.test.js:
```diff
@@ -0,0 +1,102 @@
+
+	afterAll( async () => {
+		await shopper.block.emptyCart();
+		await page.goto( 'http://localhost:8889/?teardown_terms_and_privacy' );
```
Instead of `http://localhost:8889/`, we can use `BASE_URL`.

tarunvijwani commented:
Sorry, I just noticed it is under draft mode. Please ignore my previous review comments. I'll delete them. 😁

tarhi-saad committed:
Fix filling wrong type of form error

With a virtual product we shouldn't need to fill the shipping details
We need to fill the billing instead

tarhi-saad committed:
Fix undefined 'termsCheckbox' bug

tarhi-saad committed:
Fix terms checkbox name

nielslange committed:
Merge branch 'add/5793-merchant-checkout-terms.test.js' of github.com:woocommerce/woocommerce-gutenberg-products-block into add/5793-merchant-checkout-terms.test.js

nielslange committed:
Merge branch 'trunk' into add/5793-merchant-checkout-terms.test.js

tarunvijwani commented on .wp-env.json:
```diff
@@ -3,6 +3,7 @@
 	"plugins": [
 		"https://downloads.wordpress.org/plugin/woocommerce.latest-stable.zip",
 		"https://github.com/WP-API/Basic-Auth/archive/master.zip",
+		"https://github.com/nielslange/woo-test-helper/archive/refs/heads/trunk.zip",
```
Thank you Niels for working on this! I understand that this plugin creates and configures the privacy and terms and conditions pages. Would it be better if we could use WordPress/e2e-test-utils to create a new page and assigns it as the privacy and terms and conditions pages? If so, we will no longer be dependent on this external plugin. Please let me know what you think.

nielslange replied:
I created the external plugin, as @wordPress/e2e-test-utils was failing to set up the desired pages, see pca54o-3k7-p2. That said, I could add the external plugin as a mock, similar to what we did with test themes in https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/ed47edc16758f533302cfff5e01cb4078889d0a5/.wp-env.json#L8-L12

That way, the plugin code becomes part of this repo and no longer requires the external plugin which lives in my personal GitHub repo.

nielslange replied:
Just converted the external plugin into a mock.

tarunvijwani commented on tests/e2e/specs/merchant/checkout-terms.test.js:
```diff
@@ -0,0 +1,105 @@
+	selectBlockByName,
+	saveOrPublish,
+} from '@woocommerce/blocks-test-utils';
+
```
```suggestion
import { createNewPost } from '@wordpress/e2e-test-utils';

```

We can import createNewPost from '@wordpress/e2e-test-utils' to create a new page.

nielslange replied:
Sure, we could do that. But creating the post is only one out of multiple steps. That's why I created the external plugin, that you mentioned before.

tarunvijwani commented on tests/e2e/specs/merchant/checkout-terms.test.js:
```diff
@@ -0,0 +1,105 @@
+
+describe( 'Merchant → Checkout → Can adjust T&S and Privacy Policy options', () => {
+	beforeAll( async () => {
+		await shopper.goToShop();
+		await page.goto( 'http://localhost:8889/?setup_terms_and_privacy' );
+		await expect( page ).toMatch( 'Terms & Privacy pages set up.' );
+		await shopper.block.emptyCart();
```
```suggestion
		//Create a new page with a title: Terms and Conditions page
		await createNewPost( {
			postType: 'page',
			title: 'Terms and Conditions page',
		} );
		await saveOrPublish();
		//Navigate to WooCommerce > Settings > Advanced  to set up 'Terms and Conditions page'
		await merchant.openSettings( 'advanced' );
```

The above code will create a new page and browse to WooCommerce > Settings > Advanced tab, where we can choose our page and save the settings.


Similarly, the privacy policy page is already present. We just need to save it on the `/wp-admin/options-privacy.php` page.

With the above method, we eliminate the dependency on an external plugin. Please correct me if my approach is incorrect.

nielslange replied:
See previous comment. 😀

tarunvijwani commented on tests/e2e/specs/merchant/checkout-terms.test.js:
```diff
@@ -0,0 +1,105 @@
+		await expect( page ).toMatch( 'Terms & Privacy pages teared down.' );
+	} );
+
+	it.skip( 'Merchant can see T&S and Privacy Policy links without checkbox', async () => {
```
I think you missed removing the skip from this test.

nielslange replied:
Good catch! I, indeed, forgot to remove that part.

tarunvijwani commented on tests/e2e/specs/merchant/checkout-terms.test.js:
```diff
@@ -0,0 +1,105 @@
+
+	afterAll( async () => {
+		await shopper.block.emptyCart();
+		await page.goto( 'http://localhost:8889/?teardown_terms_and_privacy' );
```
```suggestion
		await page.goto( `${ process.env.WORDPRESS_BASE_URL }/?teardown_terms_and_privacy` );
```
We can use our base URL instead (if we want to use the plugin to create pages for the test)

nielslange replied:
Also, a good catch. I replaced the static URL with a dynamic one. To prevent that the line breaks, I added the new variable `export const BASE_URL = config.get( 'url' );` to `/tests/utils/constants.js`. That way, the command fits in one line.

tarunvijwani commented on tests/e2e/specs/merchant/checkout-terms.test.js:
```diff
@@ -0,0 +1,105 @@
+describe( 'Merchant → Checkout → Can adjust T&S and Privacy Policy options', () => {
+	beforeAll( async () => {
+		await shopper.goToShop();
+		await page.goto( 'http://localhost:8889/?setup_terms_and_privacy' );
```
```suggestion
		await page.goto( `${ process.env.WORDPRESS_BASE_URL }/?setup_terms_and_privacy` );
```
We can use our base URL instead (if we want to use the plugin to create pages for the test)

nielslange replied:
See previous comment.

nielslange committed:
Activate test case

nielslange committed:
Replace static URL with dynamic one

nielslange committed:
Convert external plugin into mock

tarunvijwani reviewed (approved):
Thank you for working on the changes. It works like a charm! Let's 🚢 it.

nielslange committed:
Fix JS linting issue
